장난감 가게 구현

BEOMKING opened the pull request:
## Summary
지난 주차에 진행했던 장난감 가게를 더 심화시키는 것이 목표입니다.

## Detail
### 초기 시작 상태
- 로그인시 토큰 발급
- 유저 등록
- 유저 삭제 (관리자 권한)
- 상품 등록 (유저 권한)

### 진행할 것

BEOMKING committed:
7주차 과제 내용으로 변경

클론 받은 프로젝트의 내용을 7주차에서 진행하던 내용으로 변경했습니다.
그동안 진행했던 것을 날리고 다시 구현하기에는 비용이 크다고 생각했습니다.

BEOMKING committed:
시큐리티 부분 수정

Commit:
역할에 따라 네이밍 변경

Commit:
유저 조회 기능 구현

Commit:
유저가 존재하지 않는 경우

johngrib commented on app/src/main/java/com/codesoom/assignment/application/ProductService.java:
```diff
@@ -1,60 +1,16 @@
+import com.codesoom.assignment.dto.ProductInquiryInfo;
+import org.springframework.security.core.Authentication;
+
+public interface ProductService {
+    /**
+     * 상품을 생성하고 리턴한다.
+     *
+     * @param data           상품 정보
+     * @param authentication 유저 정보
+     * @return 상품
+     */
+    ProductInquiryInfo register(ProductData data, Authentication authentication);
```
하나의 메소드만 갖고 있는 인터페이스이므로 `ProductRegisterService`라는 이름으로 바꾸는 것도 생각할 수 있을 것 같습니다. 만약 기능을(method를) 계속 추가해 갈 것이라면 확장 범위를 제한할 목적으로 `ProductCommandService`라고 이름을 짓는 것도 생각할 수 있겠죠. 만약 `ProductCommandService`라는 이름을 사용하게 된다면 대칭적으로 `ProductQueryService`를 만들어서 전자는 데이터를 업데이트하고 추가하고 삭제하는 기능으로, 후자는 조회만 하는 기능으로 제한해 사용할 수 있을 겁니다.

BEOMKING replied:
DB에 영향을 주는 기능들과 조회 기능을 분리해서 좋은 결과를 얻었던 기억이 나네요!

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/Product.java:
```diff
@@ -1,35 +1,47 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
     private Long id;
 
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "user_id")
+    private User user;
```
이 `user`는 어떤 의미를 갖고 있나요? 상품을 구매한 사람인가요? 아니면 상품을 생산한 사람인가요? 또는 상품을 판매하는 사람인가요? 아니면 `user`라는 이름 그대로 상품을 사용하는 사람인가요?

BEOMKING replied:
유저의 의미가 모호하네요! 이 유저는 상품을 등록한 사람 즉, 판매자라고 할 수 있습니다.

owner 로 이름을 변경하면 좋을 것 같아요. seller도 괜찮을 것 같네요.

johngrib replied:
만약 각기 다른 판매자가 같은 상품을 판매하고 있다면 어떨까요?

BEOMKING replied:
음,, 질문을 잘 이해하지 못했습니다.

같은 상품을 서로 다른 판매자가 판다면 다른 상품으로 봐야할 것 같은데.. A가 파는 제로콜라와 B가 파는 제로콜라가 다른 상품인 것처럼요.

같은 상품을 여러 판매자가 팔 수 있다는 것이면 `user` 한명이 아닌 자료구조가 필요해보입니다.

johngrib replied:
네 다른 상품으로 볼 수 있죠. 이건 편의상 쇼핑몰마다 @BEOMKING 님처럼 접근하는 경우가 많습니다. (정답이 없는 문제입니다)
그런데 여기에 물류창고나 재고 개념이 들어가면 미묘해지는 경우가 있어요.
예를 들어 물류창고에 판매자 A의 제로콜라가 100개 있고 판매자 B의 제로콜라가 200개 있는 상황 같은 거죠. 누구의 제로콜라인지와 별개로 물류창고에서 상품을 카운트하기 위한 기준도 생각하게 됩니다. 여기에서 나오는 개념이 SKU 같은 거고요. (한번 조사해 보세요)

BEOMKING replied:
`SKU(Stock Keeping Unit)`는 재고관리의 최소 단위라고 합니다.
상품의 생산지, 용량 같은 상품의 정보에 따라서도 SKU가 달라질 수 있다고 합니다. 그래서 기업의 복잡성을 표시하는 대표적인 지표로 언급이 된다네요.

상품코드와 SKU, 판매자 정보로 분류할 수 있다면 상품을 카운트하는데 도움이 될 것 같은데 정말 복잡해지겠네요!
[참고](https://m.blog.naver.com/PostView.naver?isHttpsRedirect=true&blogId=kjhoi79&logNo=220593020054)

johngrib replied:
거기에 추가로 재고의 수량이 들어가면 복잡해지죠. 그리고 물류센터가 1개가 아니라 여러개가 되면 물류센터별 sku 입고 수량도 관리해야 해서 굉장히 복잡해집니다. 다양하게 조사하고 훑어보시기라도 하면 커머스와 관련된 간접 경험을 얻으실 수 있을 거에요.

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/Product.java:
```diff
@@ -1,35 +1,47 @@
-    private String maker;
+    private String description;
+
+    private Integer quantity;
```
`Integer`와 `int`의 사용을 두고 고민해 볼 필요가 있어 보입니다. `Integer`는 nullable 하죠.

BEOMKING replied:
수량을 나타내는 `quantity`의 경우에는 `int`를 사용하더라도 기본 값이 0이 되어 문제가 없을 것 같아요. 0개이면 판매할 수 없도록 기능이 되어있을테니까요.

반면에 가격을 나타내는 `price`의 경우에는 기본값으로 0이 들어가면 문제가 있을 것 같아요. 원하지 않게 0원으로 판매되는 상황이 발생할 수도 있지 않을까 생각합니다.

johngrib replied:
한 가지 시도해볼만한 과제는 `Quantity` 클래스를 만들어 `int` 대신 적용해보는 것입니다. 여유가 있다면 한번 시도해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/User.java:
```diff
@@ -1,52 +1,68 @@
+     * @param password 비밀번호
+     * @return 같으면 true, 같지 않거나 null 값이 있다면 false
+     */
+    public boolean isMatchPassword(String password) {
+        return Objects.equals(this.password, password);
```
두 가지 생각을 해볼 수 있겠어요.

1. 이 메소드가 `User`의 책임인지 고민해 볼 필요가 있습니다.
2. 만약 패스워드가 암호화되어 저장된다고 가정해 봅시다. (여기에서 암호화에 대해서는 복잡하게 생각하지 않고, 그냥 패스워드를 간단한 규칙으로 변형시키는 정도라고 생각하도록 하죠) 그렇다면 `isMatchPassword`의 구현은 어떻게 바뀌어야 할까요?

BEOMKING replied:
1. 암호화를 고려한다면 이 비교는 암호화하는 객체의 책임이 될 것 같아요.
2. 받은 패스워드와 저장되어있는 유저의 패스워드를 보내고 유저의 패스워드를 복호화해서 비교한 결과를 리턴할 것 같아요.

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/UserRepository.java:
```diff
@@ -1,15 +1,49 @@
-    Optional<User> findById(Long id);
-
-    Optional<User> findByIdAndDeletedIsFalse(Long id);
+    /**
+     * 모든 유저를 삭제합니다.
+     *
+     * 테스트에서 사용합니다.
+     */
+    void deleteAll();
```
테스트에서만 사용하는 메소드라면 `/test` 경로에서 `UserRepository` 인터페이스를 상속하거나 구현해서 만든 다른 클래스에 정의해서 사용하는 방법도 생각할 수 있을 겁니다.

BEOMKING replied:
테스트에서만 사용하는 기능을 구현해야하나 고민했었는데 이런 방법이 있었군요!

johngrib commented on app/src/main/java/com/codesoom/assignment/dto/UserRegisterData.java:
```diff
@@ -0,0 +1,25 @@
+
+@Getter
+public class UserRegisterData {
+    private final String email;
+    private final String password;
+    private final String name;
```
만약 `email`, `password`, `name` 세 가지에 대한 getter 메소드만을 갖는 인터페이스를 만든다고 생각해 봅시다. 이 인터페이스의 이름을 어떻게 지어주면 적절할까요? 그리고 그 인터페이스를 잘 사용한다면 `toUser` 메소드를 삭제할 수도 있을 겁니다.

BEOMKING replied:
유저 등록 요청에 필요한 행위를 정의한 인터페이스 `UserRegisterRequest` 를 생성했습니다. [a7706e3](https://github.com/CodeSoom/spring-week8-assignment-1/pull/65/commits/a7706e3630b4e99acb75c12d3f680e3259de5522)

더 개선할 방법이 있을까요?

johngrib replied:
오 충분히 잘 사용하신 것 같습니다.

BEOMKING committed:
모호했던 user-> 주인이라는 의미를 가진 owner로 변경

BEOMKING committed:
서비스 역할에 따라 구분

BEOMKING committed:
수량을 의미하는 Quantity 타입 Integer -> int

BEOMKING committed:
암호화 테스트 작성

BEOMKING committed:
유저 패스워드 암호화 적용

johngrib commented on app/src/main/java/com/codesoom/assignment/dto/ProductInquiryInfo.java:
```diff
@@ -26,7 +26,7 @@ public ProductInquiryInfo(Long id, Long userId, String name, String description,
+                .userId(product.getOwner().getId())
                 .name(product.getName())
                 .description(product.getDescription())
                 .quantity(product.getQuantity())
```
`ProductInquiryInfo.builder`가 `id`, `userId` 등을 일일이 지정하는 방식이 아니라 그냥 `product`를 받게 하면 어떨까요? 이것이 가능할까요? 가능하지 않거나 어렵다면 무엇 때문일까요?

BEOMKING replied:
e7f49ff
빌더 패턴을 구현해서 만들어보았습니다. 가능한 것 같습니다.
그런데 이렇게 사용했더니  `new ProductInquiryInfo(product)` 과 차이가 없는 것 같습니다.

BEOMKING replied:
> ProductInquiryInfo.builder가 id, userId 등을 일일이 지정하는 방식이 아니라 그냥 product를 받게 하면 어떨까요? 이것이 가능할까요? 가능하지 않거나 어렵다면 무엇 때문일까요?
> [e7f49ff](https://github.com/CodeSoom/spring-week8-assignment-1/commit/e7f49ff648feea503d566da40841aefc2ad69434) 빌더 패턴을 구현해서 만들어보았습니다. 가능한 것 같습니다. 그런데 이렇게 사용했더니 `new ProductInquiryInfo(product)` 과 차이가 없는 것 같습니다.

빌더 패턴에 각 필드를 받는 것이 아닌 객체 자체를 받으면 어떤 장점이 있나요?

BEOMKING replied:
> `ProductInquiryInfo.builder`가 `id`, `userId` 등을 일일이 지정하는 방식이 아니라 그냥 `product`를 받게 하면 어떨까요? 이것이 가능할까요? 가능하지 않거나 어렵다면 무엇 때문일까요?

ㅇ

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/Product.java:
```diff
@@ -26,7 +26,7 @@ public class Product {
     private String description;
 
-    private Integer quantity;
+    private int quantity;
```
이제 `null`은 방지되었지만 마이너스 수량이 가능하다는 문제는 여전히 남아 있습니다.

BEOMKING replied:
위에서 멘트 남겨주신 `Quantity` 클래스를 만들어보면서 해결할 수 있을 것 것 같은 느낌이 옵니다!

johngrib replied:
바로 캐치하셨네요 ㅎㅎ

johngrib commented on app/src/test/java/com/codesoom/assignment/application/UserCommandServiceTest.java:
```diff
@@ -0,0 +1,93 @@
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+@SpringBootTest
+@DisplayName("UserService 인터페이스의")
```
인터페이스라는 것을 굳이 명시해주지 않아도 됩니다. 인터페이스라는 건 참고사항일 뿐이죠. 만약 `UserService`를 클래스로 선언했다 하더라도 이 테스트의 내용은 모두 참이어야 합니다.

```suggestion
@DisplayName("UserService의")
```

johngrib commented on app/src/main/java/com/codesoom/assignment/security/UserAuthentication.java:
```diff
@@ -5,20 +5,28 @@
     }
 
+    private static List<GrantedAuthority> authorities(Role role) {
+        List<GrantedAuthority> authorities = new ArrayList<>();
```
이렇게 하면 어떤 차이점이 있을까요?

```suggestion
        List<GrantedAuthority> authorities = new ArrayList<>(1);
```

BEOMKING replied:
기존 방법은 개수를 예측할 수 없는 리스트 자료구조를 보는 듯 하고
제안하신 방법은 리스트 자료구조를 가지고 있지만 들어가는 요소가 1개라고 명시적으로 알려주는 느낌입니다.

johngrib replied:
그렇습니다. 그리고 한 가지 더 생각해볼 점이 있습니다. `ArrayList`의 생성자를 읽어보면서 생각해 보세요.

BEOMKING replied:
크기를 할당하지 않으면 초기 크기가 10인 `ArrayList`가 생성됩니다. 
(실제로는 크기가 0인 배열이 할당되는데 값을 추가하면 크기가 증가하는 로직이 있어 10으로 생성된다는 주석을 적은 것 같습니다.

크기를 할당한다면 해당 크기를 가진 `ArrayList`가 생성됩니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/errors/ControllerErrorAdvice.java:
```diff
@@ -0,0 +1,58 @@
+import java.util.Set;
+
+@RestControllerAdvice
+public class ControllerErrorAdvice {
```
주제별로 각 메소드를 다른 클래스로 분배하는 것도 괜찮아 보입니다.

BEOMKING replied:
연관되지 않은 예외들이 너무 몰려있는 느낌이 있었습니다. 훨씬 보기 좋아지겠네요!

BEOMKING committed:
Update app/src/test/java/com/codesoom/assignment/application/UserCommandServiceTest.java

Co-authored-by: JohnGrib <johngrib82@gmail.com>

BEOMKING committed:
Quantity 클래스 구현

BEOMKING committed:
Quantity 객체 Product에 적용

BEOMKING committed:
Merge branch 'BEOMKING' of https://github.com/BJP-Learning-Space/spring-week8-assignment-1 into BEOMKING

BEOMKING commented on app/src/main/java/com/codesoom/assignment/domain/Product.java:
```diff
@@ -26,10 +28,15 @@ public class Product {
 
     private Integer price;
 
+    public int getQuantity() {
+        return quantity.getQuantity();
+    }
```
상품을 상품 조회 DTO로 변환하는 과정에서 `product.getQuantity().getQuantity()` 라는 코드가 나와서 위처럼 변경했는데 적절한 사용일까요?

객체의 뎁스가 늘어날수록 이 getter 메소드가 계속 하위의 getter 메소드를  호출하게 될 것 같습니다. `.get().get().get()` 이 늘어나는 것보다는 나을 것 같기도 하고..흠,,

johngrib replied:
Quantity 클래스를 어떻게 디자인하느냐에 따라 달라지기도 합니다. 만약 Quantity 를 불변 데이터 객체로 설계했다면 사실 `get()`은 필요없을 수도 있습니다.

BEOMKING committed:
Update app/src/main/java/com/codesoom/assignment/security/UserAuthentication.java

Co-authored-by: JohnGrib <johngrib82@gmail.com>

BEOMKING committed:
ControllerAdvice 분리

BEOMKING committed:
상품 가격 래핑

johngrib commented on app/src/main/java/com/codesoom/assignment/errors/SessionControllerAdvice.java:
```diff
@@ -0,0 +1,25 @@
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+
+/**
+ * 세션과 연관해서 생기는 예외를 처리합니다.
+ */
+@RestControllerAdvice
+public class SessionControllerAdvice {
```
좋습니다. 한편 `Advice`라는 단어가 어째서 컴퓨터/프로그래밍에서 사용되게 되었는지에 대해 함께 읽어볼만한 글이 있습니다.

https://johngrib.github.io/wiki/clipping/marvin-minsky/why-programming-is-a-good-medium/

BEOMKING replied:
`ControllerAdvice`는 애플리케이션의 결함들에 대한 조언을 의미하는 건가요?

BEOMKING committed:
Price 테스트 실패하는 것 수정

BEOMKING committed:
UserRegisterData가 UserRegisterRequest를 상속

BEOMKING committed:
상품 수정 테스트 준비

BEOMKING commented on app/src/test/java/com/codesoom/assignment/controllers/ProductControllerTest.java:
```diff
@@ -112,10 +106,36 @@ void prepare() throws Exception {
+        class Context_without_authorization {
+            Long productId;
+
+            @BeforeEach
+            void prepare() {
+                User user = userRepository.save(Fixture.USER);
+                Product product = productRepository.save(Fixture.makeProduct(user));
+                productId = product.getId();
+            }
```
기존에는 직접 컨트롤러에 생성 요청을 보내는 것으로 테스트를 준비했습니다.
```java
private <T> Map<String, String> postRequest(Map<String, T> data, String path) throws Exception {
        return objectMapper.readValue(mockMvc.perform(post(path)
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(data)))
                .andReturn()
                .getResponse()
                .getContentAsString(), new TypeReference<>() {
        });
    }
```
이런 방법으로요. 그런데 토큰과 패스워드 암호화 같은 것이 생기면서 실제 생성 요청을 보내서 테스트를 준비하는 것이 힘들어졌어요. 

`Repository`, `Service`를 사용해서 테스트를 준비하는 방법은 컨트롤러를 테스트할 때는 없어야 하지 않을까 생각했는데 어쩔 수 없이 사용하게 됬네요.

종립님은 테스트를 준비할 때 어떤 방법을 사용하시나요?

johngrib replied:
저도 종종 이렇게 합니다. 테스트를 준비하는 공간과 테스트를 실행하는 공간의 역할 구분을 열심히 생각해 보시면 감이 잡힐 거에요. 이건 말하자면 영화 촬영을 준비하는 공간과 영화 촬영의 차이와도 비슷합니다. 멋있는 영화가 만들어지게 하기 위해 카메라 뒤편에서는 온갖 일이 일어나죠. 가능한 한 다른 레이어의 도움을 적게 받는 것도 중요한 일이지만, 테스트에서 가장 중요한 것은 검증을 위한 논리이고, 논리를 튼튼하게 쌓기 위해 테스트를 준비할 때에도 다양한 수단을 사용할 수 있다는 걸 꼭 기억해두세요.

BEOMKING committed:
불필요한 의존성 제거

BEOMKING committed:
상품 토큰이 주어지지 않는 경우

BEOMKING committed:
ProductCommandService의 update 구현

BEOMKING committed:
상품 컨트롤러 업데이트 구현

johngrib commented on app/src/main/java/com/codesoom/assignment/application/NormalProductCommandService.java:
```diff
@@ -31,4 +33,21 @@ public ProductInquiryInfo register(ProductData data, Authentication authenticati
+
+    @Override
+    public Product update(Long productId, ProductData productData, Authentication authentication) {
+        Long userId = (Long) authentication.getPrincipal();
```
이런 건 `final` 로 선언해주는 것도 괜찮겠죠.

BEOMKING committed:
상품 판매자가 아니라도 관리자라면 수정 가능하다

BEOMKING committed:
상품의 판매자도 관리자도 아니라면 수정 불가

BEOMKING committed:
변하지 않는 인스턴스 final 세팅

BEOMKING committed:
책임에 따라 메서드 분리, @Transactional 추가

메서드를 분리하면서 메서드 범위를 넘어서면 hibernate session이 종료되어
다른 메서드에서 fetch lazy 할 수 없던 문제 발생
애노테이션을 추가해서 같은 트랜잭션으로 묶음

BEOMKING committed:
상품을 찾지 못하면 발생하는 예외 핸들링

BEOMKING committed:
유저 조회, 유저 생성 메서드 시그니처 변경

상품에 유저를 할당하는 과정에서 유저 정보 조회를 다시 유저로
만들어주는 불필요한 상황이 계속 발생해서 테스트에 용이하게 변경

johngrib commented on app/src/main/java/com/codesoom/assignment/application/NormalProductCommandService.java:
```diff
@@ -44,7 +45,7 @@ public Product update(Long productId, ProductData productData, Authentication au
                 .orElseThrow(() -> new ProductNotFoundException(productId));
 
-        if (!userId.equals(product.getOwner().getId())) {
+        if (!userId.equals(product.getOwner().getId()) && authentication.getCredentials().equals(Role.USER)) {
```
여기에 생각해볼 문제가 하나 더 있습니다. 관리자도 다 똑같은 관리자가 아니라 레벨이 있을 수 있다는 거요. 레벨이 낮다면 관리할 수 있는 일이 몇 가지 있고, 레벨이 높을수록 낮은 레벨이 할 수 있는 일을 모두 할 수 있는 경우를 생각해 볼 수 있겠죠.

johngrib commented on app/src/main/java/com/codesoom/assignment/application/NormalProductCommandService.java:
```diff
@@ -38,17 +42,25 @@ public ProductInquiryInfo register(ProductData data, Authentication authenticati
+    }
 
-        return product.update(productData);
+    private void checkIfUserExists(Long userId) {
```
`checkIfUserExists` 이 이름을 `thrrowIfUserExists` 또는 `assertUserExists` 처럼 변경하는 것도 고려해 보세요.

BEOMKING committed:
빌더에 Product 를 넣는 구조로 변경

BEOMKING committed:
유저 존재 확인 메서드 네이밍 변경

BEOMKING committed:
주문 도메인 생성, 구매자와 판매자는 같을 수 없다.

BEOMKING committed:
주문 수량이 상품 수량보다 많으면 예외를 던진다.

예외가 도메인에 종속적일 수 있도록 도메인 내에 정의했습니다.
RuntimeException을 확장하고 있기에 예외명에서 Exception을 뺐습니다.

BEOMKING committed:
클래스 패키지 이동 및 관계 설정

BEOMKING committed:
주문 생성 테스트 작성

johngrib reviewed (approved):
LGTM

BEOMKING commented on app/src/main/java/com/codesoom/assignment/dto/ProductInquiryInfo.java:
```diff
@@ -23,14 +21,32 @@ public ProductInquiryInfo(Long id, Long userId, String name, String description,
+
+        ProductInquiryInfoBuilder() {}
+
+        public ProductInquiryInfoBuilder product(final Product product) {
+            this.id = product.getId();
+            this.userId = product.getOwner().getId();
+            this.name = product.getName();
+            this.description = product.getDescription();
+            this.quantity = product.getQuantity();
+            this.price = product.getPrice();
+            return this;
+        }
```
https://github.com/CodeSoom/spring-week8-assignment-1/pull/65#discussion_r952531240

빌더에 각 필드를 받는 형식이 아닌 상품 자체를 받는 방법으로 바꿨을 때 new ProductInquiryInfo(product) 과 차이가 없는 것 같습니다. 어떤 장점이 있나요??

johngrib replied:
지금 상태라면 차이가 없죠. 한 가지 더 시도해보는 것도 가치가 있겠습니다. `Product`를 has a 관계로 갖고 있게 하는 방법이죠. `Product`를 받아서 생성자에서 각 필드를 할당해주는 게 아니라 `Product product`를 그대로 간직하고 `getter`에서 `product`의 각 `getter`를 불러주게 하면 됩니다. 이걸 깔끔하게 하려고 하다 보면 여러가지 바리에이션을 시도해볼 수 있죠.

BEOMKING committed:
주문 필드의 유저가 잘못 네이밍 되어있던 것 수정

주문 필드의 유저는 seller가 아니라 buyer

BEOMKING committed:
주문 생성 구현

BEOMKING committed:
주문 생성 요청 API 구현

BEOMKING committed:
생성자 중복 제거?

BEOMKING commented on app/src/main/java/com/codesoom/assignment/domain/Order.java:
```diff
@@ -40,12 +40,8 @@ public Order(User buyer, Product product, int quantity) {
     public Order(Long id, User buyer, Product product, int quantity) {
-        throwIfBuyerAndSellerIsSame(buyer, product);
-        throwIfPurchaseQuantityGreaterThanProductStock(product, quantity);
+        this(buyer, product, quantity);
```
필드 개수에 따라서 여러 생성자가 만들어질 때 중복된 코드를 줄이기 위해 생성자가 생성자를 호출하는 방법을 사용하는 건 어떤가요?

johngrib replied:
그런 방법도 선택지중의 하나로 올려두면 좋죠. 전 이런 경우에는 생성자 함수의 이름이 꽤 아쉬워서 그냥 적절한 이름을 갖는 정적 팩토리 메소드를 만들곤 합니다.

BEOMKING commented:
8주가 벌써 끝났네요. 그동안 감사했습니다. 많은 것을 배우고 갑니다 ㅎㅎ
종립님 건강하셨으면 좋겠습니다!!

johngrib commented:
> 8주가 벌써 끝났네요. 그동안 감사했습니다. 많은 것을 배우고 갑니다 ㅎㅎ 종립님 건강하셨으면 좋겠습니다!!

고생 많으셨습니다! 코드숨 과정이 끝나도 종종 연락 주고 받도록 합시다. 트위터나 이메일로 소식 전해주세요.
